qcoro/qml/qcoroqmltask.h

```diff
@@ -0,0 +1,67 @@
+// SPDX-FileCopyrightText: 2022 Jonah Brüchert <jbb@kaidan.im>
+//
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <QCoro/QCoroTask>
+#include <QJSValue>
+#include <QJSEngine>
+
+#include "qcoroqml_export.h"
+
+#include <type_traits>
+
+namespace QCoro {
+
+//! QML type that allows to react to asynchronous computations from QML
+struct QCOROQML_EXPORT QmlTask {
+    Q_GADGET
+
+public:
+    // Just for Q_DECLARE_METATYPE to be happy
+    QmlTask() noexcept;
+
+    //! Constructs a QmlTask from a QCoro::Task that resolves to a QVariant
+    /*!
+     * \param[in] task to await
+     */
+    QmlTask(QCoro::Task<QVariant> &&task);
+
+    //! Constructs a QmlTask from a QCoro::Task that resolves to an arbitrary type
+    /*!
+     * \param[in] task to await
+     */
+    template <typename T>
+    requires std::is_convertible_v<T, QVariant>
+    QmlTask(QCoro::Task<T> &&task) : QmlTask(
+        task.then([](T result) -> QCoro::Task<QVariant> {
+            co_return QVariant(result);
+        }))
+    {
+    }
+
+    template <typename T = void>
+    QmlTask(QCoro::Task<> &&task) : QmlTask(
+        task.then([]() -> QCoro::Task<QVariant> {
+            co_return QVariant();
+        }))
+    {
+    }
+
+    //! QML function that executes a given JavaScript function once the computation is finished
+    /*!
+     * \param[in] JavaScript function to call once the result is ready.
+     *  The result will be passed as first argument to the given function
+     */
+    Q_INVOKABLE void then(QJSValue func);
+
+private:
+    // To allow the object to be copyable (required by Q_DECLARE_METATYPE) without actually copying
+    // Since the whole object has the size of a single pointer, we can potentially swap this for a d-ptr if needed.
```

Comment:
Shared pointer (both `std` and Qt) are actually size of two pointers -  (unlike `QSharedDataPointer` which indeed has size of one pointer).

I'm inclined to play it safe and use a d-pointer here.

Reply:
Do you know if there is a smart pointer that just always copies the contained object?
Doing another round of ref-counting with a QSharedDataPointer seems unnecessary.

Reply:
I'm not aware of any such smart pointer; but if the the d-pointer is a QSharedDataPointer, then the `m_task` no longer has to be a shared pointer as well, does it? One just has to be careful to not call a non-const QSharedDataPointer method that would cause a detach (and add an assert to QmlTaskPrivate's copy ctor), then you get the same behavior, right?

Reply:
A right, I didn't expect this to work, but it actually does. Nice